graph/graphs/gen/gen.go

```diff
@@ -54,106 +54,150 @@ func (s IDSet) Len() int       { return len(s) }
+	switch path.Len() {
+	case 0:
+		return
+	case 1:
 		u, new := dst.NodeWithID(path.ID(0))
 		if new {
 			dst.AddNode(u)
 		}
-		return nil
+		return
+	}
+	err := check(path)
+	if err != nil {
+		panic(err)
 	}
 	for i := 0; i < path.Len()-1; i++ {
 		uid := path.ID(i)
 		vid := path.ID(i + 1)
-		if uid == vid {
-			return fmt.Errorf("gen: adjacent node IDs equal at %d: %d", i, uid)
-		}
 		u, _ := dst.NodeWithID(uid)
 		v, _ := dst.NodeWithID(vid)
 		dst.SetEdge(dst.NewEdge(u, v))
 	}
-	return nil
 }
 
 // Star constructs a star graph in dst with edges between the center node ID to
 // node with IDs specified in leaves.
 // If dst is a directed graph, edges are directed from the center node to the
-// leaves.
-func Star(dst NodeIDGraphBuilder, center int64, leaves IDer) error {
+// leaves. If any ID appears twice in leaves and center, Star will panic.
+func Star(dst NodeIDGraphBuilder, center int64, leaves IDer) {
 	c, new := dst.NodeWithID(center)
 	if new {
 		dst.AddNode(c)
 	}
+	if leaves.Len() == 0 {
+		return
+	}
+	err := check(leaves, center)
+	if err != nil {
+		panic(err)
+	}
 	for i := 0; i < leaves.Len(); i++ {
 		id := leaves.ID(i)
-		if id == center {
-			return fmt.Errorf("gen: leaf %d ID matches central node ID: %d", i, center)
-		}
 		n, _ := dst.NodeWithID(id)
 		dst.SetEdge(dst.NewEdge(c, n))
 	}
-	return nil
 }
 
 // Wheel constructs a wheel graph in dst with edges from the center
 // node ID to node with IDs specified in cycle and between nodes with IDs
 // adjacent in the cycle.
 // If dst is a directed graph, edges are directed from the center node to the
-// cycle and from earlier nodes to later nodes in cycle.
-func Wheel(dst NodeIDGraphBuilder, center int64, cycle IDer) error {
-	err := Star(dst, center, cycle)
-	if err != nil {
-		return err
+// cycle and from earlier nodes to later nodes in cycle. If any ID appears
+// twice in cycle and center, Wheel will panic.
+func Wheel(dst NodeIDGraphBuilder, center int64, cycle IDer) {
+	Star(dst, center, cycle)
+	if cycle.Len() <= 1 {
+		return
+	}
+	cycleNoCheck(dst, cycle)
+}
+
+// check confirms that no node ID exists more than once in ids and extra.
+func check(ids IDer, extra ...int64) error {
+	seen := make(map[int64]int, ids.Len()+len(extra))
+	for j := 0; j < ids.Len(); j++ {
+		uid := ids.ID(j)
+		if prev, exists := seen[uid]; exists {
+			return fmt.Errorf("gen: node ID collision i=%d j=%d: id=%d", prev, j, uid)
+		}
+		seen[uid] = j
+	}
+	lenIDs := ids.Len()
+	for j, uid := range extra {
+		if prev, exists := seen[uid]; exists {
+			if prev < lenIDs {
+				if len(extra) == 1 {
+					return fmt.Errorf("gen: node ID collision i=%d with extra: id=%d", prev, uid)
+				}
+				return fmt.Errorf("gen: node ID collision i=%d with extra j=%d: id=%d", prev, j, uid)
+			}
+			prev -= lenIDs
+			return fmt.Errorf("gen: extra node ID collision i=%d j=%d: id=%d", prev, j, uid)
+		}
+		seen[uid] = j + lenIDs
 	}
-	return Cycle(dst, cycle)
+	return nil
 }
```

Comment:
Should there be some way to let the programmer disable checking? Supposing they have already checked the input, calling `check` would be redundant. A boolean parameter could be added to the graph-building functions to achieve this, although this would modify their signatures.

Reply:
I thought about that and if they are confident enough, I'd rather they copy/paste the code that does the work in the actual routines here than add a knob. These are convenience operations and the costs here are small compared to the actual graph routines being performed in most cases.